server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -0,0 +1,165 @@
+                    {
+                        builder.startObject("synonyms");
+                        {
+                            builder.field("type", "text");
+                        }
+                        builder.endObject();
+                        builder.startObject(SYNONYMS_SET_FIELD);
+                        {
+                            builder.field("type", "keyword");
+                        }
+                        builder.endObject();
+                    }
+                    builder.endObject();
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to build mappings for " + SYNONYMS_INDEX, e);
+        }
+    }
+
+    public void putSynonymSet(String resourceName, SynonymSet synonymSet, ActionListener<PutSynonymsAction.Response> listener) {
+
+        // TODO Add synonym rules validation
+
+        // Delete synonym set if it existed previously
+        DeleteByQueryRequest dbqRequest = new DeleteByQueryRequest(SYNONYMS_INDEX).setQuery(
+            QueryBuilders.termQuery(SYNONYMS_SET_FIELD, resourceName)
+        ).setIndicesOptions(IndicesOptions.fromOptions(true, true, false, false));
+
+        client.execute(
+            DeleteByQueryAction.INSTANCE,
+            dbqRequest,
+            listener.delegateFailure((deleteByQueryResponseListener, bulkByScrollResponse) -> {
+                boolean created = bulkByScrollResponse.getDeleted() == 0;
+
+                // Insert as bulk requests
+                BulkRequestBuilder bulkRequestBuilder = client.prepareBulk();
+                try {
+                    for (SynonymRule synonymRule : synonymSet.synonyms()) {
+
+                        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                            builder.startObject();
+                            {
+                                builder.field(SYNONYMS_FIELD, synonymRule.synonym());
+                                builder.field(SYNONYMS_SET_FIELD, resourceName);
+                            }
+                            builder.endObject();
+
+                            final IndexRequest indexRequest = new IndexRequest(SYNONYMS_INDEX).opType(DocWriteRequest.OpType.INDEX)
+                                .source(builder);
+                            final String synonymRuleId = synonymRule.id();
+                            if (synonymRuleId != null) {
+                                indexRequest.id(synonymRuleId);
+                            }
+
+                            bulkRequestBuilder.add(indexRequest);
+                        }
+                    }
+                } catch (IOException ex) {
+                    listener.onFailure(ex);
+                }
+
+                bulkRequestBuilder.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
```

Comment:
Is it possible that chaining these two requests could result in a race condition, where you're updating the synonym set with one request and deleting it with another request?

Reply:
Yes, it is. Same as with Search Applications approach.

Reply:
That's my concern - perhaps we could find a better path forward here, and apply a similar solution to search applications?

Reply:
Same as happened in Search Apps, I don't think this is a blocker as of now. We can think of a better approach in follow up work.

Reply:
Will this be launching as tech preview as well?

Reply:
My assumption is that this request - update the whole synonyms set  will be very rare, so it is ok if there is race condition.  An alternative to use `synchronize` should slow things a lot, to my mind.

Reply:
We may at least want to create a followup issue to fix.